File: scripts/scaffold/steps/replacements.py
"""Steps 5-6: Replace repository references and language namespaces."""

from __future__ import annotations

import os

from ..config import Config
from ..files import (
    is_git_dir,
    rename_dotnet_files,
    rename_java_dirs,
    replace_in_dockerfiles,
    replace_in_file,
    replace_in_tree,
)
from ..log import fatal, log, ok

# All text file extensions to process
TEXT_EXTS = [
    ".yml", ".yaml", ".md", ".gradle", ".gradle.kts",
    ".csproj", ".sln", ".slnx", ".cshtml", ".json",
    ".cs", ".java", ".ts", ".tsx", ".js", ".jsx",
    ".xml", ".properties", ".cfg", ".txt",
]


# ─── Step 5: Replace repository references ──────────────────────────────────


def replace_repo_references(cfg: Config, **_: object) -> None:
    log("Step 5: Replacing repository references...")

    if cfg.dry_run:
        log(f"[DRY RUN] Would replace optivem/starter -> {cfg.full_repo}")
        return

    repo_dir = cfg.repo_dir

    # Pass 1: optivem/starter -> owner/repo
    n = replace_in_tree(repo_dir, "optivem/starter", cfg.full_repo, TEXT_EXTS)
    n += replace_in_dockerfiles(repo_dir, "optivem/starter", cfg.full_repo)
    ok(f"Pass 1: replaced optivem/starter -> {cfg.full_repo} ({n} files)")

    # Pass 2: optivem_starter -> owner_repo (SonarCloud underscore variant)
    underscore_new = f"{cfg.owner}_{cfg.repo}"
    n = replace_in_tree(repo_dir, "optivem_starter", underscore_new, TEXT_EXTS)
    ok(f"Pass 2: replaced optivem_starter -> {underscore_new} ({n} files)")

    # Pass 3: SonarCloud org (scoped patterns to avoid touching optivem/actions)
    sonar_replacements = [
        ("'sonar.organization', 'optivem'", f"'sonar.organization', '{cfg.owner_lower}'"),
        ('/o:"optivem"', f'/o:"{cfg.owner_lower}"'),
        ("-Dsonar.organization=optivem", f"-Dsonar.organization={cfg.owner_lower}"),
    ]
    for old, new in sonar_replacements:
        n = replace_in_tree(repo_dir, old, new)
        if n > 0:
            ok(f"Pass 3: replaced sonar org pattern ({n} files)")

    # Safety check: optivem/actions must still be intact
    wf_dir = os.path.join(repo_dir, ".github", "workflows")
    actions_found = any(
        "optivem/actions" in open(os.path.join(wf_dir, f), encoding="utf-8").read()
        for f in os.listdir(wf_dir)
        if f.endswith(".yml")
    ) if os.path.isdir(wf_dir) else False

    if not actions_found:
        fatal("Safety check failed: optivem/actions references were corrupted during replacement!")
    ok("Safety check passed: optivem/actions references intact")

    _lowercase_docker_compose_images(repo_dir)


def _lowercase_docker_compose_images(repo_dir: str) -> None:
    for dirpath, _dirnames, filenames in os.walk(repo_dir):
        if is_git_dir(dirpath):
            continue
        for fname in filenames:
            if "docker-compose" not in fname or not fname.endswith(".yml"):
                continue
            filepath = os.path.join(dirpath, fname)
            try:
                with open(filepath, "r", encoding="utf-8") as f:
                    lines = f.readlines()
                changed = False
                for i, line in enumerate(lines):
                    if "image:" in line and "ghcr.io" in line:
                        prefix, _, rest = line.partition("image:")
                        lowered = f"{prefix}image:{rest.lower()}"
                        if lowered != lines[i]:
                            lines[i] = lowered
                            changed = True
                if changed:
                    with open(filepath, "w", encoding="utf-8", newline="\n") as f:
                        f.writelines(lines)
            except (OSError, UnicodeDecodeError):
                pass
    ok("Docker-compose image URLs lowercased")


# ─── Step 6: Replace namespaces ─────────────────────────────────────────────


def replace_namespaces(cfg: Config, **_: object) -> None:
    log("Step 6: Replacing namespaces...")

    if cfg.dry_run:
        log("[DRY RUN] Would replace language-specific namespaces")
        return

    if cfg.arch == "monolith":
        assert cfg.lang is not None
        _ns_for_lang(cfg, cfg.lang, "monolith")
        _ns_for_lang(cfg, cfg.test_lang, "systemtest")
    else:
        assert cfg.backend_lang is not None
        _ns_for_lang(cfg, cfg.backend_lang, "backend")
        _ns_for_lang(cfg, cfg.test_lang, "systemtest")

    ok("Namespace replacement complete")


def _ns_for_lang(cfg: Config, lang: str, component: str) -> None:
    if lang == "java":
        _ns_java(cfg, component)
    elif lang == "dotnet":
        _ns_dotnet(cfg, component)
    elif lang == "typescript":
        _ns_typescript(cfg, component)


def _ns_java(cfg: Config, component: str) -> None:
    old_full = f"{cfg.java_ns_old}.{component}"
    new_full = f"{cfg.java_ns_new}.{component}"
    repo_dir = cfg.repo_dir

    n = replace_in_tree(repo_dir, old_full, new_full, [".java", ".gradle", ".gradle.kts", ".xml", ".properties"])
    n += replace_in_tree(repo_dir, old_full, new_full, [".yml"])
    ok(f"Java: replaced {old_full} -> {new_full} ({n} files)")

    old_dir_parts = ["com", "optivem", "starter"]
    new_dir_parts = ["com", cfg.owner_lower, cfg.repo_nohyphens]
    for dirpath, _dirnames, _filenames in os.walk(repo_dir):
        if is_git_dir(dirpath):
            continue
        if os.path.isdir(os.path.join(dirpath, *old_dir_parts)):
            rename_java_dirs(dirpath, old_dir_parts, new_dir_parts)
    ok(f"Java: renamed directories com/optivem/starter -> com/{cfg.owner_lower}/{cfg.repo_nohyphens}")


def _ns_dotnet(cfg: Config, component: str) -> None:
    component_map = {"monolith": "Monolith", "backend": "Backend", "systemtest": "SystemTest"}
    old_full = f"{cfg.dotnet_ns_old}.{component_map[component]}"
    new_full = f"{cfg.dotnet_ns_new}.{component_map[component]}"
    repo_dir = cfg.repo_dir

    n = replace_in_tree(repo_dir, old_full, new_full, [".cs", ".cshtml", ".csproj", ".sln", ".slnx", ".json", ".yml"])
    n += replace_in_dockerfiles(repo_dir, old_full, new_full)
    ok(f".NET: replaced {old_full} -> {new_full} ({n} files)")

    rename_dotnet_files(repo_dir, old_full, new_full)
    ok(f".NET: renamed files {old_full}.* -> {new_full}.*")


def _ns_typescript(cfg: Config, component: str) -> None:
    if component != "systemtest":
        return

    repo_dir = cfg.repo_dir
    n = replace_in_tree(repo_dir, cfg.ts_pkg_old, cfg.ts_pkg_new, [".json"])
    ok(f"TypeScript: replaced {cfg.ts_pkg_old} -> {cfg.ts_pkg_new} ({n} files)")

    for dirpath, _dirnames, filenames in os.walk(repo_dir):
        if "system-test" in dirpath and "package.json" in filenames:
            pkg_path = os.path.join(dirpath, "package.json")
            replace_in_file(pkg_path, '"author": "Optivem"', f'"author": "{cfg.owner}"')
            replace_in_file(pkg_path, '"Starter - System Tests"', f'"{cfg.system_name} - System Tests"')
            replace_in_file(pkg_path, '"optivem"', f'"{cfg.owner_lower}"')
            ok("TypeScript: updated package.json metadata")
            break

    for dirpath, _dirnames, filenames in os.walk(repo_dir):
        if is_git_dir(dirpath) or "system-test" in dirpath or "node_modules" in dirpath:
            continue
        if "package.json" in filenames:
            pkg_path = os.path.join(dirpath, "package.json")
            if "monolith" in dirpath:
                replace_in_file(pkg_path, '"name": "starter-monolith"', f'"name": "{cfg.repo}-monolith"')
            elif "backend" in dirpath:
                replace_in_file(pkg_path, '"name": "starter-backend"', f'"name": "{cfg.repo}-backend"')

File: gh-optivem/internal/steps/replacements.go
package steps

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/optivem/gh-optivem/internal/config"
	"github.com/optivem/gh-optivem/internal/files"
	"github.com/optivem/gh-optivem/internal/log"
	"github.com/optivem/gh-optivem/internal/templates"
)

// All text file extensions to process.
var textExts = []string{
	".yml", ".yaml", ".md", ".gradle", ".gradle.kts",
	".csproj", ".sln", ".slnx", ".cshtml", ".json",
	".cs", ".java", ".ts", ".tsx", ".js", ".jsx",
	".xml", ".properties", ".cfg", ".txt",
}

// ReplaceRepoReferences replaces optivem/starter references with the target repo.
func ReplaceRepoReferences(cfg *config.Config) {
	log.Log("Step 6: Replacing repository references...")

	if cfg.DryRun {
		log.Logf("[DRY RUN] Would replace optivem/starter -> %s", cfg.FullRepo)
		return
	}

	if cfg.Arch == "monolith" {
		replaceRefsInRepo(cfg.RepoDir, cfg.FullRepo, cfg.OwnerLower)
	} else {
		replaceRefsInRepo(cfg.RepoDir, cfg.FullRepo, cfg.OwnerLower)
		replaceRefsInRepo(cfg.BackendRepoDir, cfg.BackendFullRepo, cfg.OwnerLower)
		replaceRefsInRepo(cfg.FrontendRepoDir, cfg.FrontendFullRepo, cfg.OwnerLower)

		// Fix docker-compose image URLs
		templates.FixupMultirepoDockerCompose(
			cfg.RepoDir, cfg.Repo, cfg.FrontendRepo, cfg.BackendRepo, cfg.BackendLang,
		)
	}

	log.OK("Repository reference replacement complete")
}

func replaceRefsInRepo(repoDir, fullRepo, ownerLower string) {
	// Pass 1: optivem/starter -> owner/repo
	n := files.ReplaceInTree(repoDir, "optivem/starter", fullRepo, textExts)
	n += files.ReplaceInDockerfiles(repoDir, "optivem/starter", fullRepo)
	log.OKf("Pass 1: replaced optivem/starter -> %s (%d files)", fullRepo, n)

	// Pass 2: optivem_starter -> owner_repo (SonarCloud underscore variant)
	underscoreNew := strings.ReplaceAll(fullRepo, "/", "_")
	n = files.ReplaceInTree(repoDir, "optivem_starter", underscoreNew, textExts)
	log.OKf("Pass 2: replaced optivem_starter -> %s (%d files)", underscoreNew, n)

	// Pass 3: SonarCloud org patterns
	sonarReplacements := [][2]string{
		{"'sonar.organization', 'optivem'", "'sonar.organization', '" + ownerLower + "'"},
		{`/o:"optivem"`, `/o:"` + ownerLower + `"`},
		{"-Dsonar.organization=optivem", "-Dsonar.organization=" + ownerLower},
	}
	for _, pair := range sonarReplacements {
		n = files.ReplaceInTree(repoDir, pair[0], pair[1], nil)
		if n > 0 {
			log.OKf("Pass 3: replaced sonar org pattern (%d files)", n)
		}
	}

	// Safety check: optivem/actions must still be intact
	wfDir := filepath.Join(repoDir, ".github", "workflows")
	if info, err := os.Stat(wfDir); err == nil && info.IsDir() {
		actionsFound := false
		entries, _ := os.ReadDir(wfDir)
		for _, e := range entries {
			if !strings.HasSuffix(e.Name(), ".yml") {
				continue
			}
			data, err := os.ReadFile(filepath.Join(wfDir, e.Name()))
			if err != nil {
				continue
			}
			if strings.Contains(string(data), "optivem/actions") {
				actionsFound = true
				break
			}
		}
		if !actionsFound {
			log.Fatalf("Safety check failed: optivem/actions references were corrupted in %s!", repoDir)
		}
		log.OKf("Safety check passed: optivem/actions references intact in %s", repoDir)
	}

	lowercaseDockerComposeImages(repoDir)
}

func lowercaseDockerComposeImages(repoDir string) {
	filepath.Walk(repoDir, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		if files.IsGitDir(path) {
			return nil
		}
		if !strings.Contains(info.Name(), "docker-compose") || !strings.HasSuffix(info.Name(), ".yml") {
			return nil
		}

		data, readErr := os.ReadFile(path)
		if readErr != nil {
			return nil
		}

		lines := strings.Split(string(data), "\n")
		changed := false
		for i, line := range lines {
			if strings.Contains(line, "image:") && strings.Contains(line, "ghcr.io") {
				idx := strings.Index(line, "image:")
				prefix := line[:idx+6]
				rest := line[idx+6:]
				lowered := prefix + strings.ToLower(rest)
				if lowered != lines[i] {
					lines[i] = lowered
					changed = true
				}
			}
		}
		if changed {
			os.WriteFile(path, []byte(strings.Join(lines, "\n")), 0644)
		}
		return nil
	})
	log.OK("Docker-compose image URLs lowercased")
}

// ReplaceNamespaces replaces language-specific namespaces.
func ReplaceNamespaces(cfg *config.Config) {
	log.Log("Step 7: Replacing namespaces...")

	if cfg.DryRun {
		log.Log("[DRY RUN] Would replace language-specific namespaces")
		return
	}

	if cfg.Arch == "monolith" {
		nsForLang(cfg, cfg.Lang, "monolith", cfg.RepoDir)
		nsForLang(cfg, cfg.TestLang, "systemtest", cfg.RepoDir)
	} else {
		// System repo: only system-test namespaces
		nsForLang(cfg, cfg.TestLang, "systemtest", cfg.RepoDir)
		// Backend repo
		nsForLang(cfg, cfg.BackendLang, "backend", cfg.BackendRepoDir)
		// Frontend repo
		if cfg.FrontendLang == "react" {
			fixupFrontendPackageJSON(cfg)
		}
	}

	log.OK("Namespace replacement complete")
}

func nsForLang(cfg *config.Config, lang, component, repoDir string) {
	switch lang {
	case "java":
		nsJava(cfg, component, repoDir)
	case "dotnet":
		nsDotnet(cfg, component, repoDir)
	case "typescript":
		nsTypeScript(cfg, component, repoDir)
	}
}

func nsJava(cfg *config.Config, component, repoDir string) {
	oldFull := cfg.JavaNsOld + "." + component
	newFull := cfg.JavaNsNew + "." + component

	n := files.ReplaceInTree(repoDir, oldFull, newFull, []string{".java", ".gradle", ".gradle.kts", ".xml", ".properties"})
	n += files.ReplaceInTree(repoDir, oldFull, newFull, []string{".yml"})
	log.OKf("Java: replaced %s -> %s (%d files)", oldFull, newFull, n)

	oldDirParts := []string{"com", "optivem", "starter"}
	newDirParts := []string{"com", cfg.OwnerLower, cfg.RepoNoHyphens}

	filepath.Walk(repoDir, func(path string, info os.FileInfo, err error) error {
		if err != nil || !info.IsDir() || files.IsGitDir(path) {
			return nil
		}
		check := filepath.Join(path, "com", "optivem", "starter")
		if _, err := os.Stat(check); err == nil {
			files.RenameJavaDirs(path, oldDirParts, newDirParts)
			return filepath.SkipDir
		}
		return nil
	})
	log.OKf("Java: renamed directories com/optivem/starter -> com/%s/%s", cfg.OwnerLower, cfg.RepoNoHyphens)
}

func nsDotnet(cfg *config.Config, component, repoDir string) {
	componentMap := map[string]string{
		"monolith": "Monolith", "backend": "Backend", "systemtest": "SystemTest",
	}
	oldFull := cfg.DotnetNsOld + "." + componentMap[component]
	newFull := cfg.DotnetNsNew + "." + componentMap[component]

	n := files.ReplaceInTree(repoDir, oldFull, newFull, []string{".cs", ".cshtml", ".csproj", ".sln", ".slnx", ".json", ".yml"})
	n += files.ReplaceInDockerfiles(repoDir, oldFull, newFull)
	log.OKf(".NET: replaced %s -> %s (%d files)", oldFull, newFull, n)

	files.RenameDotnetFiles(repoDir, oldFull, newFull)
	log.OKf(".NET: renamed files %s.* -> %s.*", oldFull, newFull)
}

func nsTypeScript(cfg *config.Config, component, repoDir string) {
	if component != "systemtest" {
		return
	}

	n := files.ReplaceInTree(repoDir, cfg.TsPkgOld, cfg.TsPkgNew, []string{".json"})
	log.OKf("TypeScript: replaced %s -> %s (%d files)", cfg.TsPkgOld, cfg.TsPkgNew, n)

	// Update package.json metadata in system-test
	filepath.Walk(repoDir, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		if strings.Contains(path, "system-test") && info.Name() == "package.json" {
			files.ReplaceInFile(path, `"author": "Optivem"`, `"author": "`+cfg.Owner+`"`)
			files.ReplaceInFile(path, `"Starter - System Tests"`, `"`+cfg.SystemName+` - System Tests"`)
			files.ReplaceInFile(path, `"optivem"`, `"`+cfg.OwnerLower+`"`)
			log.OK("TypeScript: updated package.json metadata")
			return filepath.SkipAll
		}
		return nil
	})

	// Update package.json in system dirs (monolith/backend)
	filepath.Walk(repoDir, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() || files.IsGitDir(path) {
			return nil
		}
		if strings.Contains(path, "system-test") || strings.Contains(path, "node_modules") {
			return nil
		}
		if info.Name() == "package.json" {
			if strings.Contains(path, "monolith") {
				files.ReplaceInFile(path, `"name": "starter-monolith"`, `"name": "`+cfg.Repo+`-monolith"`)
			} else if strings.Contains(path, "backend") {
				files.ReplaceInFile(path, `"name": "starter-backend"`, `"name": "`+cfg.Repo+`-backend"`)
			}
		}
		return nil
	})
}

func fixupFrontendPackageJSON(cfg *config.Config) {
	pkgPath := filepath.Join(cfg.FrontendRepoDir, "package.json")
	if _, err := os.Stat(pkgPath); err == nil {
		files.ReplaceInFile(pkgPath, `"name": "starter-frontend"`, `"name": "`+cfg.Repo+`-frontend"`)
		files.ReplaceInFile(pkgPath, `"name": "frontend-react"`, `"name": "`+cfg.Repo+`-frontend"`)
	}
}
